File: beta v1.0.0/Dev Planner.app/Contents/Resources/src/core/task_node.py
from PyQt6.QtWidgets import QWidget, QVBoxLayout, QHBoxLayout, QLabel, QPushButton, QLineEdit, QTextEdit, QMenu, QGraphicsDropShadowEffect
from PyQt6.QtCore import Qt, QPoint, QPointF, QRectF, QTimer, QPropertyAnimation, QEasingCurve
from PyQt6.QtGui import QPainter, QPainterPath, QColor, QPen, QPixmap, QIcon
from src.core.config import STATUSES


def create_color_icon(color_hex, size=16):
    pixmap = QPixmap(size, size)
    pixmap.fill(Qt.GlobalColor.transparent)
    painter = QPainter(pixmap)
    painter.setRenderHint(QPainter.RenderHint.Antialiasing)
    painter.setBrush(QColor(color_hex))
    painter.setPen(Qt.PenStyle.NoPen)
    painter.drawEllipse(2, 2, size - 4, size - 4)
    painter.end()
    return QIcon(pixmap)


class TaskNode(QWidget):
    BASE_WIDTH = 220
    BASE_HEIGHT = 140
    
    def __init__(self, x, y, canvas, title="Новая задача"):
        super().__init__(canvas)
        self.canvas = canvas
        self.node_x = x
        self.node_y = y
        self.status = 'none'
        self.connections = []
        self.dragging = False
        self.drag_offset = QPoint()
        self.connecting = False
        self.is_hover_target = False
        
        self.setFixedSize(self.BASE_WIDTH, self.BASE_HEIGHT)
        self.setMouseTracking(True)
        self.move(int(x), int(y))
        self.setStyleSheet("background: transparent;")
        
        self._setup_ui(title)
        self.show()
    
    def update_scale(self, scale):
        new_width = int(self.BASE_WIDTH * scale)
        new_height = int(self.BASE_HEIGHT * scale)
        self.setFixedSize(new_width, new_height)
    
    def _setup_ui(self, title):
        self.main_layout = QVBoxLayout(self)
        self.main_layout.setContentsMargins(12, 10, 12, 10)
        self.main_layout.setSpacing(6)
        
        header = QHBoxLayout()
        header.setSpacing(8)
        
        self.status_indicator = QLabel()
        self.status_indicator.setFixedSize(12, 12)
        self.update_status_indicator()
        
        self.title_edit = QLineEdit(title)
        self.title_edit.setContextMenuPolicy(Qt.ContextMenuPolicy.NoContextMenu)
        self.title_edit.setStyleSheet("""
            QLineEdit {
                background: transparent;
                border: none;
                outline: none;
                color: #ffffff;
                font-size: 13px;
                font-weight: bold;
                selection-background-color: rgba(0,255,255,0.3);
            }
            QLineEdit:focus {
                border: none;
                outline: none;
            }
        """)
        self.title_edit.textChanged.connect(self.on_change)

        self.delete_btn = QPushButton("×")
        self.delete_btn.setFixedSize(20, 20)
        self.delete_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        self.delete_btn.setStyleSheet("""
            QPushButton { background: transparent; color: #ff0055; border: none; font-size: 16px; font-weight: bold; }
            QPushButton:hover { color: #ffffff; }
        """)
        self.delete_btn.clicked.connect(self.delete_self)
        
        header.addWidget(self.status_indicator)
        header.addWidget(self.title_edit, 1)
        header.addWidget(self.delete_btn)
        
        self.desc_edit = QTextEdit()
        self.desc_edit.setPlaceholderText("Описание...")
        self.desc_edit.setContextMenuPolicy(Qt.ContextMenuPolicy.NoContextMenu)
        self.desc_edit.setStyleSheet("""
            QTextEdit {
                background: rgba(0,0,0,0.3);
                border: none;
                border-radius: 6px;
                color: #cccccc;
                font-size: 11px;
                padding: 5px;
                selection-background-color: rgba(0,255,255,0.3);
            }
            QTextEdit:focus {
                border: none;
                outline: none;
            }
        """)
        self.desc_edit.textChanged.connect(self.on_change)
        
        self.title_edit.installEventFilter(self)
        self.desc_edit.installEventFilter(self)
        self.desc_edit.viewport().installEventFilter(self)
        self.delete_btn.installEventFilter(self)
        
        self.main_layout.addLayout(header)
        self.main_layout.addWidget(self.desc_edit, 1)
    
    def eventFilter(self, obj, event):
        if self.canvas.connecting_from:
            if event.type() in (
                event.Type.MouseButtonPress,
                event.Type.MouseButtonRelease,
                event.Type.MouseButtonDblClick,
                event.Type.FocusIn,
                event.Type.KeyPress,
            ):
                if event.type() == event.Type.MouseButtonPress:
                    self.mousePressEvent(event)
                return True
        return super().eventFilter(obj, event)
    
    def on_change(self):
        if hasattr(self.canvas, 'main_window') and self.canvas.main_window:
            self.canvas.main_window.schedule_autosave()
    
    def update_status_indicator(self):
        color = STATUSES[self.status]['color']
        self.status_indicator.setStyleSheet(f"""
            background-color: {color};
            border-radius: 6px;
            border: 2px solid {color};
        """)
    
    def set_status(self, status):
        self.status = status
        self.update_status_indicator()
        self.canvas.update()
        self.on_change()
    
    def delete_self(self):
        self.canvas.remove_node(self)
    
    def get_center(self):
        return QPointF(self.node_x + self.BASE_WIDTH/2, self.node_y + self.BASE_HEIGHT/2)

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing)
        
        color = QColor(STATUSES[self.status]['color'])
        path = QPainterPath()
        path.addRoundedRect(QRectF(0, 0, self.width(), self.height()), 10, 10)
        
        if self.is_hover_target:
            painter.fillPath(path, QColor(0, 255, 255, 40))
            painter.setPen(QPen(QColor("#00ffff"), 3))
        else:
            painter.fillPath(path, QColor(20, 20, 20, 240))
            painter.setPen(QPen(color, 2))
        painter.drawPath(path)
        painter.end()
    
    def set_hover_target(self, is_target):
        if self.is_hover_target != is_target:
            self.is_hover_target = is_target
            self.update()
    
    def mousePressEvent(self, event):
        if self.canvas.connecting_from:
            self.set_hover_target(False)
            if self.canvas.connecting_from != self:
                self.canvas.complete_connection(self)
            else:
                self.canvas.cancel_connection()
            event.accept()
            return
            
        if event.button() == Qt.MouseButton.LeftButton:
            self.dragging = True
            self.drag_offset = event.pos()
            self.raise_()
        elif event.button() == Qt.MouseButton.RightButton:
            self.show_context_menu(event.globalPosition().toPoint())
    
    def enterEvent(self, event):
        if self.canvas.connecting_from and self.canvas.connecting_from != self:
            self.set_hover_target(True)
            self.canvas.hover_target = self
    
    def leaveEvent(self, event):
        if self.is_hover_target:
            self.set_hover_target(False)
            if self.canvas.hover_target == self:
                self.canvas.hover_target = None
    
    def mouseMoveEvent(self, event):
        if self.canvas.connecting_from:
            global_pos = self.mapToParent(event.pos())
            self.canvas.mouse_pos = QPointF(global_pos.x(), global_pos.y())
            self.canvas.connection_overlay.update()
            return
            
        if self.dragging:
            new_pos = self.mapToParent(event.pos() - self.drag_offset)
            self.move(new_pos)
            self.node_x = (new_pos.x() - self.canvas.offset.x()) / self.canvas.scale
            self.node_y = (new_pos.y() - self.canvas.offset.y()) / self.canvas.scale
            self.canvas.update()
            self.canvas.connection_overlay.update()
    
    def mouseReleaseEvent(self, event):
        self.dragging = False
        self.on_change()
    
    def show_context_menu(self, pos):
        menu = QMenu(self)
        menu.setStyleSheet("""
            QMenu { background-color: #1a1a1a; border: 1px solid #333333; border-radius: 5px; padding: 5px; }
            QMenu::item { color: #ffffff; padding: 8px 20px; border-radius: 3px; }
            QMenu::item:selected { background-color: #333333; }
        """)
        
        status_menu = menu.addMenu("Статус")
        status_menu.setStyleSheet("""
            QMenu { background-color: #1a1a1a; border: 1px solid #333333; border-radius: 5px; padding: 5px; }
            QMenu::item { color: #ffffff; padding: 8px 20px; border-radius: 3px; }
            QMenu::item:selected { background-color: #333333; }
        """)
        for key, value in STATUSES.items():
            color = value['color']
            icon = create_color_icon(color)
            action = status_menu.addAction(icon, value['name'])
            action.triggered.connect(lambda checked, k=key: self.set_status(k))
        
        menu.addSeparator()
        connect_action = menu.addAction("Соединить с...")
        connect_action.triggered.connect(self.start_connection)
        menu.addSeparator()
        delete_action = menu.addAction("Удалить")
        delete_action.triggered.connect(self.delete_self)
        menu.exec(pos)
    
    def start_connection(self):
        self.canvas.start_connection(self)
    
    def get_data(self):
        return {
            'title': self.title_edit.text(),
            'description': self.desc_edit.toPlainText(),
            'status': self.status,
            'x': self.node_x,
            'y': self.node_y
        }
    
    def load_data(self, data):
        self.title_edit.setText(data.get('title', 'Новая задача'))
        self.desc_edit.setPlainText(data.get('description', ''))
        self.status = data.get('status', 'none')
        self.update_status_indicator()

File: cpp/src/ui/task_node.cpp
#include "task_node.hpp"
#include "core/config.hpp"
#include "node_canvas.hpp"
#include <QHBoxLayout>
#include <QMenu>
#include <QMouseEvent>
#include <QPainter>
#include <QPainterPath>
#include <QPixmap>
#include <QVBoxLayout>

namespace DevPlanner {

TaskNode::TaskNode(qreal x, qreal y, NodeCanvas *canvas, const QString &title,
                   QWidget *parent)
    : GlassmorphismWidget(parent), m_canvas(canvas), m_nodeX(x), m_nodeY(y) {
  setFixedSize(BASE_WIDTH, BASE_HEIGHT);
  setMouseTracking(true);
  setAttribute(Qt::WA_TranslucentBackground);
  move(static_cast<int>(x), static_cast<int>(y));
  setupUI(title);
  show();
}

void TaskNode::setupUI(const QString &title) {
  auto *layout = new QVBoxLayout(this);
  layout->setContentsMargins(15, 12, 15, 12);
  layout->setSpacing(8);

  auto *header = new QHBoxLayout();
  header->setSpacing(10);

  m_statusIndicator = new QLabel(this);
  updateStatusIndicator();

  m_titleEdit = new QLineEdit(title, this);
  m_titleEdit->setContextMenuPolicy(Qt::NoContextMenu);
  m_titleEdit->setStyleSheet(
      "QLineEdit { background: transparent; border: none; color: #ffffff; "
      "font-size: 14px; font-weight: 800; }");
  connect(m_titleEdit, &QLineEdit::textChanged, this,
          &TaskNode::onContentChanged);
  m_titleEdit->installEventFilter(this);

  m_deleteBtn = new QPushButton("×", this);
  m_deleteBtn->setFixedSize(24, 24);
  m_deleteBtn->setStyleSheet(
      "QPushButton { background: transparent; color: rgba(255,255,255,0.3); "
      "border: none; font-size: 20px; } QPushButton:hover { color: #ff0055; }");
  connect(m_deleteBtn, &QPushButton::clicked, this, &TaskNode::onDeleteClicked);
  m_deleteBtn->installEventFilter(this);

  header->addWidget(m_statusIndicator);
  header->addWidget(m_titleEdit, 1);
  header->addWidget(m_deleteBtn);

  m_descEdit = new QTextEdit(this);
  m_descEdit->setContextMenuPolicy(Qt::NoContextMenu);
  m_descEdit->setPlaceholderText("Notes...");
  m_descEdit->setStyleSheet(
      "QTextEdit { background: rgba(0,0,0,0.2); border: 1px solid "
      "rgba(255,255,255,0.05); "
      "border-radius: 12px; color: rgba(255,255,255,0.8); font-size: 13px; "
      "padding: 10px; }");
  connect(m_descEdit, &QTextEdit::textChanged, this,
          &TaskNode::onContentChanged);
  m_descEdit->installEventFilter(this);
  m_descEdit->viewport()->installEventFilter(this);

  if (title.isEmpty()) {
    m_titleEdit->hide();
    layout->setContentsMargins(10, 8, 10, 8);
    layout->setSpacing(4);
    auto *miniHeader = new QHBoxLayout();
    miniHeader->addWidget(m_statusIndicator);
    miniHeader->addStretch();
    miniHeader->addWidget(m_deleteBtn);
    layout->addLayout(miniHeader);
  } else {
    layout->addLayout(header);
  }
  layout->addWidget(m_descEdit, 1);
}

bool TaskNode::eventFilter(QObject *obj, QEvent *event) {
  if (m_canvas && m_canvas->isConnecting()) {
    if (event->type() == QEvent::MouseButtonPress) {
      auto *me = static_cast<QMouseEvent *>(event);
      mousePressEvent(me);
      return true;
    }
  }
  return GlassmorphismWidget::eventFilter(obj, event);
}

void TaskNode::onContentChanged() { emit changed(); }
void TaskNode::onDeleteClicked() { emit deleteRequested(this); }

void TaskNode::updateStatusIndicator() {
  QString c = getStatuses()[m_status].color;
  m_statusIndicator->setStyleSheet(
      QString("background-color: %1; border-radius: 6px; "
              "min-width: 12px; max-width: 12px; "
              "min-height: 12px; max-height: 12px;")
          .arg(c));
}

void TaskNode::setStatus(const QString &s) {
  if (getStatuses().contains(s)) {
    m_status = s;
    updateStatusIndicator();
    update();
    emit changed();
  }
}
void TaskNode::setNodePosition(qreal x, qreal y) {
  m_nodeX = x;
  m_nodeY = y;
}
QString TaskNode::title() const { return m_titleEdit->text(); }
void TaskNode::setTitle(const QString &t) { m_titleEdit->setText(t); }
QString TaskNode::description() const { return m_descEdit->toPlainText(); }
void TaskNode::setDescription(const QString &d) { m_descEdit->setPlainText(d); }
QPointF TaskNode::getCenter() const {
  return QPointF(m_nodeX + BASE_WIDTH / 2.0, m_nodeY + BASE_HEIGHT / 2.0);
}
void TaskNode::updateScale(qreal s) {
  qreal effectiveScale = qMax(s, 0.5);

  setFixedSize(static_cast<int>(BASE_WIDTH * effectiveScale),
               static_cast<int>(BASE_HEIGHT * effectiveScale));

  int titleFontSize = qMax(static_cast<int>(14 * effectiveScale), 9);
  int descFontSize = qMax(static_cast<int>(13 * effectiveScale), 8);
  int btnSize = qMax(static_cast<int>(24 * effectiveScale), 14);
  int indicatorSize = qMax(static_cast<int>(10 * effectiveScale), 6);
  int margins = qMax(static_cast<int>(15 * effectiveScale), 8);

  m_titleEdit->setStyleSheet(
      QString(
          "QLineEdit { background: transparent; border: none; color: #ffffff; "
          "font-size: %1px; font-weight: 800; }")
          .arg(titleFontSize));

  m_descEdit->setStyleSheet(
      QString("QTextEdit { background: rgba(0,0,0,0.2); border: 1px solid "
              "rgba(255,255,255,0.05); "
              "border-radius: %1px; color: rgba(255,255,255,0.8); font-size: "
              "%2px; padding: %3px; }")
          .arg(qMax(static_cast<int>(12 * effectiveScale), 6))
          .arg(descFontSize)
          .arg(qMax(static_cast<int>(10 * effectiveScale), 4)));

  m_deleteBtn->setFixedSize(btnSize, btnSize);
  m_deleteBtn->setStyleSheet(
      QString("QPushButton { background: transparent; color: "
              "rgba(255,255,255,0.3); "
              "border: none; font-size: %1px; } QPushButton:hover { color: "
              "#ff0055; }")
          .arg(qMax(static_cast<int>(20 * effectiveScale), 12)));

  m_statusIndicator->setFixedSize(indicatorSize, indicatorSize);
  QString statusColor = getStatuses()[m_status].color;
  int radius = indicatorSize / 2;
  m_statusIndicator->setStyleSheet(
      QString("background-color: %1; border-radius: %2px; "
              "min-width: %3px; max-width: %3px; "
              "min-height: %3px; max-height: %3px;")
          .arg(statusColor)
          .arg(radius)
          .arg(indicatorSize));

  if (layout()) {
    layout()->setContentsMargins(margins, margins - 3, margins, margins - 3);
  }
}
void TaskNode::setHoverTarget(bool t) {
  if (m_isHoverTarget != t) {
    m_isHoverTarget = t;
    update();
  }
}

void TaskNode::paintEvent(QPaintEvent *event) {
  Q_UNUSED(event);
  QPainter painter(this);
  painter.setRenderHint(QPainter::Antialiasing, true);

  QColor statusColor(getStatuses()[m_status].color);
  QPainterPath path;
  path.addRoundedRect(rect().adjusted(1, 1, -1, -1), 16, 16);

  painter.setClipPath(path);
  painter.fillPath(path, QColor(12, 12, 20, 250));

  painter.setClipping(false);
  QPen pen(m_isHoverTarget ? statusColor : QColor(60, 60, 80),
           m_isHoverTarget ? 2 : 1);
  painter.setPen(pen);
  painter.drawPath(path);
}

void TaskNode::mousePressEvent(QMouseEvent *e) {
  if (m_canvas && m_canvas->isConnecting()) {
    setHoverTarget(false);
    if (m_canvas->getConnectingFrom() != this)
      m_canvas->completeConnection(this);
    else
      m_canvas->cancelConnection();
    e->accept();
    return;
  }
  if (e->button() == Qt::LeftButton) {
    m_isDragging = true;
    m_dragOffset = e->pos();
    raise();
  } else if (e->button() == Qt::RightButton)
    showContextMenu(e->globalPosition().toPoint());
}

void TaskNode::mouseMoveEvent(QMouseEvent *e) {
  if (m_canvas && m_canvas->isConnecting()) {
    QPoint p = mapToParent(e->pos());
    m_canvas->updateMousePosition(QPointF(p));
    return;
  }
  if (m_isDragging) {
    QPoint p = mapToParent(e->pos() - m_dragOffset);
    move(p);
    if (m_canvas) {
      m_nodeX = (p.x() - m_canvas->offset().x()) / m_canvas->scale();
      m_nodeY = (p.y() - m_canvas->offset().y()) / m_canvas->scale();
      m_canvas->update();
      m_canvas->updateConnectionOverlay();
    }
  }
}

void TaskNode::mouseReleaseEvent(QMouseEvent *e) {
  m_isDragging = false;
  emit changed();
}
void TaskNode::enterEvent(QEnterEvent *e) {
  if (m_canvas && m_canvas->isConnecting() &&
      m_canvas->getConnectingFrom() != this) {
    setHoverTarget(true);
    m_canvas->setHoverTarget(this);
  }
}
void TaskNode::leaveEvent(QEvent *e) {
  if (m_isHoverTarget) {
    setHoverTarget(false);
    if (m_canvas)
      m_canvas->setHoverTarget(nullptr);
  }
}

void TaskNode::showContextMenu(const QPoint &p) {
  QMenu menu(this);
  menu.setStyleSheet(
      "QMenu { background: rgba(30,30,40,240); border: 1px solid "
      "rgba(255,255,255,0.1); border-radius: 10px; color: #ffffff; padding: "
      "5px; } QMenu::item:selected { background: rgba(217,0,255,0.3); }");
  auto *s = menu.addMenu("Status");
  for (auto it = getStatuses().begin(); it != getStatuses().end(); ++it) {
    auto *a = s->addAction(it.value().name);
    QString k = it.key();
    connect(a, &QAction::triggered, this, [this, k]() { setStatus(k); });
  }
  menu.addSeparator();
  connect(menu.addAction("Connect"), &QAction::triggered, this,
          [this]() { emit connectionRequested(this); });
  connect(menu.addAction("Delete"), &QAction::triggered, this,
          &TaskNode::onDeleteClicked);
  menu.exec(p);
}

void TaskNode::startConnection() { emit connectionRequested(this); }
QJsonObject TaskNode::getData() const {
  QJsonObject o;
  o["title"] = m_titleEdit->text();
  o["description"] = m_descEdit->toPlainText();
  o["status"] = m_status;
  o["x"] = m_nodeX;
  o["y"] = m_nodeY;
  return o;
}
void TaskNode::loadData(const QJsonObject &d) {
  m_titleEdit->setText(d["title"].toString());
  m_descEdit->setPlainText(d["description"].toString());
  m_status = d["status"].toString("none");
  updateStatusIndicator();
}

} // namespace DevPlanner
